go/vt/topo/etcd2topo/server_test.go

```diff
@@ -89,6 +96,117 @@ func startEtcd(t *testing.T) (*exec.Cmd, string, string) {
 	return cmd, dataDir, clientAddr
 }
 
+// startEtcdWithTLS starts an etcd subprocess with TLS setup, and waits for it to be ready.
+func startEtcdWithTLS(t *testing.T) (*exec.Cmd, string, string) {
+	// Create a temporary directory.
+	dataDir, err := ioutil.TempDir("", "etcd")
+	if err != nil {
+		t.Fatalf("cannot create tempdir: %v", err)
+	}
+
+	// Get our two ports to listen to.
+	port := testfiles.GoVtTopoEtcd2topoPort
+	name := "vitess_unit_test"
+	clientAddr := fmt.Sprintf("http://localhost:%v", port)
+	peerAddr := fmt.Sprintf("http://localhost:%v", port+1)
+	initialCluster := fmt.Sprintf("%v=%v", name, peerAddr)
+
+	// Generate cert and key in /tmp
+	err = exec.Command("openssl", "req", "-new", "-newkey", "rsa:4096", "-days", "365", "-nodes", "-x509",
+		"-subj", "/C=US/ST=Denial/L=Springfield/O=Dis/CN=127.0.0.1",
+		"-keyout", testKeyPath, "-out", testCertPath).Run()
+	if err != nil {
+		t.Fatalf("unable to generate test key and cert using openssl: %v", err)
+	}
+
+	cmd := exec.Command("etcd",
+		"-name", name,
+		"-advertise-client-urls", clientAddr,
+		"-initial-advertise-peer-urls", peerAddr,
+		"-listen-client-urls", clientAddr,
+		"-listen-peer-urls", peerAddr,
+		"-initial-cluster", initialCluster,
+		"-cert-file", testCertPath,
+		"-key-file", testKeyPath,
+		"-data-dir", dataDir)
+	err = cmd.Start()
+	if err != nil {
+		t.Fatalf("failed to start etcd: %v", err)
+	}
+
+	// Safe now to build up TLS info.
+	tlsInfo := transport.TLSInfo{
+		CertFile:      testCertPath,
+		KeyFile:       testKeyPath,
+		TrustedCAFile: testCertPath,
+	}
+
+	tlsConfig, err := tlsInfo.ClientConfig()
+
+	// Create a client to connect to the created etcd.
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{clientAddr},
+		TLS:         tlsConfig,
+		DialTimeout: 5 * time.Second,
+	})
+	if err != nil {
+		t.Fatalf("newCellClient(%v) failed: %v", clientAddr, err)
+	}
+
+	// Wait until we can list "/", or timeout.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	start := time.Now()
+	for {
+		if _, err := cli.Get(ctx, "/"); err == nil {
+			break
+		}
+		if time.Since(start) > 10*time.Second {
+			t.Fatalf("Failed to start etcd daemon in time")
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	return cmd, dataDir, clientAddr
+}
+
+func TestEtcd2TLS(t *testing.T) {
+	// Start a single etcd in the background.
+	cmd, dataDir, clientAddr := startEtcdWithTLS(t)
+	testIndex := 0
+	testRoot := fmt.Sprintf("/test-%v", testIndex)
+
+	// Create the server on the new root.
+	server, err := NewServerWithOpts(clientAddr, testRoot, testCertPath, testKeyPath, testCertPath)
+	if err != nil {
+		t.Fatalf("NewServerWithOpts failed: %v", err)
+	}
+	client := server.cli
+
+	defer func() {
+		cmd.Process.Kill()
+		cmd.Wait()
+		os.RemoveAll(dataDir)
+		server.Close()
+		client.Close()
+	}()
+
+	testCtx := context.Background()
+	testKey := "testkey"
+	testVal := "testval"
+	_, err = client.Put(testCtx, testKey, testVal)
```

Comment:
Instead of pulling the unexported `cli` field and messing with it, I'd still rather we just test the public API of `server`, since there's no need for this test to depend on implementation details to ensure the behavior it's meant to verify.

```go
_, err := server.Create(testCtx, testKey, []byte(testVal))
...
val, _, err := server.Get(testCtx, testKey)
...
```

Reply:
Done https://github.com/vitessio/vitess/pull/5496/commits/eafeeb45fac2954d4c0e05f3f4c2979bf226aa1b.